include/oneapi/dpl/internal/binary_search_impl.h

```diff
@@ -68,49 +68,57 @@ struct custom_brick
     }
 };
 
-template <typename Policy, typename InputIterator1, typename InputIterator2, typename OutputIterator,
+template <class _Tag, typename Policy, typename InputIterator1, typename InputIterator2, typename OutputIterator,
           typename StrictWeakOrdering>
-oneapi::dpl::__internal::__enable_if_host_execution_policy<Policy, OutputIterator>
-lower_bound_impl(Policy&& policy, InputIterator1 start, InputIterator1 end, InputIterator2 value_start,
+OutputIterator
+lower_bound_impl(_Tag, Policy&& policy, InputIterator1 start, InputIterator1 end, InputIterator2 value_start,
                  InputIterator2 value_end, OutputIterator result, StrictWeakOrdering comp)
 {
+    static_assert(__internal::__is_backend_tag_serial_v<_Tag>);
+
     return oneapi::dpl::transform(policy, value_start, value_end, result,
                                   [=](typename ::std::iterator_traits<InputIterator2>::reference val) {
                                       return ::std::lower_bound(start, end, val, comp) - start;
                                   });
 }
 
-template <typename Policy, typename InputIterator1, typename InputIterator2, typename OutputIterator,
+template <class _Tag, typename Policy, typename InputIterator1, typename InputIterator2, typename OutputIterator,
           typename StrictWeakOrdering>
-oneapi::dpl::__internal::__enable_if_host_execution_policy<Policy, OutputIterator>
-upper_bound_impl(Policy&& policy, InputIterator1 start, InputIterator1 end, InputIterator2 value_start,
+OutputIterator
+upper_bound_impl(_Tag, Policy&& policy, InputIterator1 start, InputIterator1 end, InputIterator2 value_start,
                  InputIterator2 value_end, OutputIterator result, StrictWeakOrdering comp)
 {
+    static_assert(__internal::__is_backend_tag_serial_v<_Tag>);
+
     return oneapi::dpl::transform(policy, value_start, value_end, result,
                                   [=](typename ::std::iterator_traits<InputIterator2>::reference val) {
                                       return ::std::upper_bound(start, end, val, comp) - start;
                                   });
 }
 
-template <typename Policy, typename InputIterator1, typename InputIterator2, typename OutputIterator,
+template <class _Tag, typename Policy, typename InputIterator1, typename InputIterator2, typename OutputIterator,
           typename StrictWeakOrdering>
-oneapi::dpl::__internal::__enable_if_host_execution_policy<Policy, OutputIterator>
-binary_search_impl(Policy&& policy, InputIterator1 start, InputIterator1 end, InputIterator2 value_start,
+OutputIterator
+binary_search_impl(_Tag, Policy&& policy, InputIterator1 start, InputIterator1 end, InputIterator2 value_start,
                    InputIterator2 value_end, OutputIterator result, StrictWeakOrdering comp)
 {
+    static_assert(__internal::__is_backend_tag_serial_v<_Tag>);
```

Comment:
@MikeDvorskiy FYI:
```C++
oneDPL/include/oneapi/dpl/internal/binary_search_impl.h:105:19: error: static assertion failed due to requirement '__internal::__is_backend_tag_serial_v<oneapi::dpl::__internal::__parallel_tag<std::integral_constant<bool, false>>>'
  105 |     static_assert(__internal::__is_backend_tag_serial_v<_Tag>);
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
yes, because an static assert (if we want to have it here) must be a more complicated  - see one of my previous comment.
"
....So, to control "dispatching" we need just one trait like is_seq_or_unseq_or_unknown_policy or/and non-random access iterators...
Other words, it is enough to check that a tag is a __serial_tag and checking expression should call like "is_serial_tag_v".... I will discuss it with Ruslan one more time."